feat(linter): add vitest/prefer-lowercase-title rule

taearls opened the pull request:
This pull request implements the [vitest/prefer-lowercase-title](https://github.com/vitest-dev/eslint-plugin-vitest/blob/main/docs/rules/prefer-lowercase-title.md) rule. 

Since there was an existing jest rule with this title, I followed the existing pattern in [no-unused-vars](https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/rules/eslint/no_unused_vars/mod.rs) to group the jest and vitest rules together in a shared module. I used the existing `jest/prefer-lowercase-title` documentation as a base and modified it where it seemed appropriate. I added a `jest` and `vitest` snapshot suffix for each respective test suite.

One item I wasn't 100% about is adding `bench` to the jest test names. Without this change, the vitest test suite fails because of [this check](https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/utils/jest/parse_jest_fn.rs#L108) which validates that we're only parsing valid jest functions from a detected jest file. The unit tests that are sourced from the vitest plugin are all read by the linting host as jest-like files, so adding `bench` as a "valid" jest method allows us to lint a unit test using this keyword. This seemed to me like the least invasive solution to accommodate the new rule without breaking any existing code, but I'm certainly open to alternatives.

taearls committed:
gen prefer lowercase title

taearls committed:
populate some of the rule logic

taearls committed:
add more logic

taearls committed:
add missing import

taearls committed:
fix import

taearls committed:
run formatter

taearls committed:
use run on jest node

taearls committed:
fix typo

taearls committed:
run rule

taearls committed:
rename file

taearls committed:
wip

taearls committed:
Merge branch 'main' into tyler/vitest-prefer-lowercase-title

taearls committed:
get closer to fix

taearls committed:
run just fix

taearls committed:
enable fail tests

taearls committed:
add debug to cargo toml

taearls committed:
Merge branch 'main' into tyler/vitest-prefer-lowercase-title

taearls committed:
explicitly get first char

taearls committed:
get tests working

taearls committed:
get all tests passing

taearls committed:
remove unused imporst

taearls committed:
code cleanup

taearls committed:
code cleanup

taearls committed:
remove dbg calls

taearls committed:
refactor

taearls committed:
remove unused import

taearls committed:
add ignores support

taearls committed:
fix lint warning

taearls committed:
undo cargo toml changes, remove added line

taearls committed:
revert cargo toml change

taearls committed:
migrate rule

taearls committed:
migrate tests

taearls committed:
code cleanup

taearls committed:
rename fn variant to bench

taearls committed:
update documentation

taearls committed:
formatting

taearls committed:
clean up formatting

taearls committed:
add bench to ignores

taearls committed:
formatting

camc314 commented on crates/oxc_linter/src/utils/vitest.rs:
```diff
@@ -19,3 +22,19 @@ pub fn parse_expect_and_typeof_vitest_fn_call<'a>(
     }
 }
+
+pub fn parse_vitest_fn_call<'a>(
+    call_expr: &'a CallExpression<'a>,
+    possible_jest_node: &PossibleJestNode<'a, '_>,
+    ctx: &LintContext<'a>,
+) -> Option<ParsedGeneralJestFnCall<'a>> {
+    let jest_fn_call = parse_jest_fn_call(call_expr, possible_jest_node, ctx)?;
+
+    match jest_fn_call {
+        ParsedJestFnCallNew::GeneralJest(jest_fn_call) => match jest_fn_call.kind {
+            JestFnKind::General(_) => Some(jest_fn_call),
+            _ => None,
+        },
+        _ => None,
+    }
+}
```
this feels misleading?

why are `ParsedJestFnCall::Expect` and `ParsedJestFnCall::ExpectTypeOf` not included?

camc314 replied:
ugh i see. github is giving me a poor diff.

i think this change should be reverted, and you should just do:
```
        let Some(ParsedJestFnCallNew::GeneralJest(jest_fn_call)) =
            parse_jest_fn_call(call_expr, possible_jest_node, ctx)
        else {
            return;
        };
```

in the rule/mod.rs

taearls replied:
done

camc314 commented on crates/oxc_linter/src/snapshots/vitest_prefer_lowercase_title2.snap:
should these be here?

same with `crates/oxc_linter/src/snapshots/vitest_prefer_lowercase_title2@vitest.snap`

taearls replied:
oops, my bad. I'll remove those.

taearls replied:
removed

camc314 commented on crates/oxc_linter/src/rules/jest/prefer_lowercase_title/mod.rs:
```diff
@@ -0,0 +1,314 @@
+
+    fn run_on_jest_node<'a, 'c>(
+        &self,
+        possible_vitest_node: &PossibleJestNode<'a, 'c>,
```
lets keep this called `possible_vitest_node`

i think the idea here is to have jest rules, adapted for vitest not the other way round

taearls replied:
just to clarify, did you mean to say that we should keep it called `possible_jest_node`?

taearls replied:
I assume that's what you meant, so I renamed it to `possible_jest_node`.

camc314 replied:
yeah that's right - thanks

taearls committed:
remove unneeded snapshots

taearls committed:
remove vitest util, update var name to possible_jest_node

taearls committed:
formatting

taearls committed:
add prefer-lowercase-title rule to VITEST_COMPATIBLE_JEST_RULES

taearls committed:
update snapshot

taearls commented:
I noticed the `VITEST_COMPATIBLE_JEST_RULES` set and added `prefer-lowercase-title` there. this also updated my snapshot test for the vitest unit tests with the correct plugin name.

camc314 reviewed (approved):
Thanks!

taearls committed:
resolve merge conflicts

taearls committed:
update snapshot

taearls commented:
I resolved the merge conflicts, so it's ready for review again.

taearls committed:
resolve merge conflicts

taearls commented:
@Boshen when you get a chance, would you mind giving this a review? thanks!
